### ISSUE TITLE
Inter-grid kernels

### ISSUE THREAD

**Issue by christophermaynard** [stfc/PSyclone#59]
Inter-grid kernels for the multigrid solver and physics/dynamics require and extension to the PSyKAl API.
This is documented in MORS Ticket 1037. The rules are repeated here, but the example is in a branch of LFRic.

* An inter-grid kernel (and metadata) must have ''at least'' one field on both the fine and coarse meshes. Specifying all fields as coarse or fine is forbidden.
* All fields must have a halo depth of at least 2. This could be a run-time check.
* Always do the horizontal looping over the coarse mesh, requiring only depth 1 halo looping
* Colouring for the coarse mesh for shared memory parallelism will be correct.
* fields on different meshes must always live on different function spaces. 
* For DG fields (W3 and W_theta) metadata for {{{any_space_dg_n}}} has been introduced so that DG looping can be employed if the fine and the coarse fields are both (for example) W3. This will allow optimisation of the horizontal looping that can be developed later.
* Only fields are allowing in inter-grid kernels. This can be re-visited later.
* Looping over the coarse cells implies the {{{whole_dof_map}}} (not just a column slice) for the ''fine'' dof_map must be passed in, along with the accompanying {{{ncell_f}}} scalar.

**Comment by arporter** [comment]
Just want to check with @christophermaynard that "fields on different meshes must always live on different function spaces" is correct? So when we prolong/restrict a field we change its function space?

**Comment by christophermaynard** [comment]
Yes, because they live on different meshes so the horizontal looping is different. Our choice of function space metadata is conflating two issues. Actually we only care about the continuity (horizontal) for the PSy layer. Other information may or may not be specific to the kernel. To get round the problem we have introduced `any_space_dg_n` so that a prolong from W3fine to W3coarse can be `any_space_dg1` to `any_space_dg2` which is actually what we mean. This is described in points 5 and 6. Tom Melvin has a comment underneath which describes the wider problem and ticket #848 has the details.

**Comment by arporter** [comment]
Created new branch `intergrid_mdata` for work to support the extensions to meta-data.

**Comment by arporter** [comment]
Branch https://code.metoffice.gov.uk/svn/lfric/LFRic/branches/dev/rickywong/r10007_FsChainUsage has examples in psykal-lite. e.g. invoke_restrict_field:

    ! get the field proxies
    fc_fp = fc%get_proxy()
    ff_fp = ff%get_proxy()

    ! get the mesh from the first field argument, this is the
    ! field which is being written too. See Kernel Metadata
    mesh => fc%get_mesh()
    ! get the map from the 1st to the 2nd , i.e. coarse to fine
    mesh_f => ff%get_mesh()
    mesh_map => mesh%get_mesh_map( mesh_f )
    ! could get nlayers from either, but lets be consistent, get it from
    ! the field we are writing too.
    nlayers = fc_fp%vspace%get_nlayers()

    ! there are multiple routes to this information which we don't want, so
    ! always get the number of cells from the mesh, not the map
    ncell_f = mesh_f%get_last_halo_cell(depth=2)
    ncell_c = mesh%get_last_halo_cell(depth=1)
    ncell_ratio = mesh_map%get_ncell_map_ratio()

    dofmap_f => ff_fp%vspace%get_whole_dofmap()
    dofmap_c => fc_fp%vspace%get_whole_dofmap()
    ndf = ff_fp%vspace%get_ndf()
    undf_f = ff_fp%vspace%get_undf()
    undf_c = fc_fp%vspace%get_undf()

    ! halo exchange to depth two on the fine as we are reading from
    if (ff_fp%is_dirty(depth=2)) then
      call ff_fp%halo_exchange(depth=2)
    end if

    ! halo exchange to depth one because we are incrementing and looping
    ! to last halo cell(1)
    if (fc_fp%is_dirty(depth=1)) then
      call fc_fp%halo_exchange(depth=1)
    end if

    cell_map => mesh_map%get_whole_cell_map()

    ! if we are using OpenMP, using the coarse grid colouring is still correct
    do cell = 1, ncell_c
      ! pass the whole dofmap for the fine (we are reading from)
      call restrict_kernel_code(nlayers, cell_map(:,cell), ncell_ratio, &
        dofmap_f, ncell_f, dofmap_c(:,cell), ndf, undf_f, undf_c, fc_fp%data, ff_fp%data)
    end do

    call fc_fp%set_dirty()

The meta-data for the kernel is:

    type(arg_type) :: meta_args(2) = (/                                     &
       arg_type(GH_FIELD, GH_INC, ANY_SPACE_1, mesh_arg=GH_COARSE),         &
       arg_type(GH_FIELD, GH_READ,  ANY_SPACE_2, mesh_arg=GH_FINE   )       &
       /)
    integer :: iterates_over = CELLS

Key difference is that we now specify which mesh each argument is on.

**Comment by arporter** [comment]
Another clarification question for @christophermaynard and/or @thomasmelvin - are field-vector arguments allowed for inter-grid kernels (e.g. `GH_FIELD*3`)?

**Comment by christophermaynard** [comment]
hmm... The 3D field (which is not a field_vector, we have field_vector_type that is a different thing). Currently are only use of field*3 is for a the coordinate field. It is not obvious that this should be part of inter-grid kernels. I think there are some ideas about having a 3-vector field (e.g. winds) but we don't have this yet. There one could transform between Physics and Dynamics which will also be inter-grid. I will talk to BeShe (who is not on github [shriek!]) tomorrow.

**Comment by arporter** [comment]
OK, thanks ChriMa.

**Comment by christophermaynard** [comment]
Oops. Fat fingers ... pressed wrong button. I can never get the hang of the PSyclon Claw things ... Thanks Andy for re-opening ...

**Comment by rupertford** [comment]
This issue is about adding support for inter-grid kernels to PSyclone.

The first phase of this support - adding appropriate metadata to PSyclone and parsing it correctly - has been completed as pr #75 is now on master.

The next phase is to add actual support for generating PSy-layer code for such kernels in PSyclone.

**Comment by arporter** [comment]
Created new `inter_grid_kernels` branch and added first test (for a prolong kernel).

**Comment by thomasmelvin** [comment]
FIELD*3 for intergrid kernels:
I've lost where this discussion is up to, but here goes...

I agree with Chris that I  dont think our current usage of these structures will be needed for the intergrid kernels, but I also dont see any reason to disallow them. Surely psyclone will just treat them the same way as it will any other field it is passed but there are 3 data arrays instead of one?

I dont think we should be ruling out there use in intergrid kernels unless we have to.

**Comment by arporter** [comment]
I'm just sorting out the rules regarding arguments to inter-grid kernels and have a further clarification question for @christophermaynard and @thomasmelvin  - if a kernel reads from more than one field on the fine mesh, are those fields permitted to be on different function spaces? If so, presumably we'll need to pass in the *whole* dofmap for each associated function space?

**Comment by arporter** [comment]
@christophermaynard and @thomasmelvin As discussed at the telco, this is the Issue for the implementation of inter-grid kernels. I think I had just begun the implementation phase by writing some tests. I'll need to bring the branch up-to-date and then press on once our funding situation is resolved. If either of you could answer the question in my previous comment that would be great.

**Comment by christophermaynard** [comment]
@arporter and @thomasmelvin. I don't think we plan to do multi-space - multi-function space operations. One, then the other if we need to do both. I would say, no, we don't need to do this. We can certainly do the "one then the other" for correctness, and then we have a nice optimisation to do! So, no we don't need an inter-grid kernels to read from multiple spaces.

**Comment by arporter** [comment]
Hi @christophermaynard, is https://code.metoffice.gov.uk/svn/lfric/LFRic/branches/dev/rickywong/r10007_FsChainUsage still the best place to be looking at what I need to generate or is it now on trunk?

**Comment by arporter** [comment]
I'm in the process of re-ordering the kernel arguments so that they are more consistent with 'normal' kernels - i.e. with dofmap info at the end of the argument list. I then need to extend `ArgOrdering` to deal with inter-grid kernels.

**Comment by arporter** [comment]
I've just re-based my branch onto master to pick up the developments there (in particular, the fixes to various Fortran test files that are needed by the latest fparser).

**Comment by BShipway** [comment]
A quick note on the relevance of this to intergrid physics kernels (following prompt by ChriMa yesterday):  

Essentially the same sort of infrastructure will be required there, except that there will also be refinement in the vertical and so nlayers (and thus the way we use the map) will be different for the coarse and fine meshes.  E.g. we will additionally want to do:
```
nlayers_f = ff_fp%vspace%get_nlayers()
nlayers_c = fc_fp%vspace%get_nlayers()
```

This requirement is still a way off, so for now this is just to be aware of and a 'do not inhibit' requirement.

**Comment by arporter** [comment]
Thanks Ben - noted!

**Comment by arporter** [comment]
Hi @christophermaynard, @thomasmelvin, just a quick question about the LFRic inter-grid API. For the restrict example in psykalite you use `get_ncell_map_ratio()` but in the prolong example you use
`get_ntarget_cells_per_source_cell()`. Looking in `mesh_map_mod.F90` I see that the former is just a pointer to the latter anyway:

    procedure, public :: get_ntarget_cells_per_source_cell
    procedure, public :: get_ncell_map_ratio => get_ntarget_cells_per_source_cell

So, I propose to use `get_ntarget_cells_per_source_cell()` in both cases. Is that OK?

**Comment by arporter** [comment]
I now have the very basic prolong and restrict examples working. Still to do:

1. Add compilation checks
2. Add support for DM (and enable this in the tests)
3. Disable transformations when loops contain inter-grid kernels
4. Add more tests for cases when we have more than one inter-grid kernel in an invoke.

**Comment by arporter** [comment]
Regarding distributed memory:

We always loop over the cells in the coarse mesh for inter-grid kernels. A halo depth of 1 on this mesh is equivalent to a halo depth of 2 on the fine mesh.

**Comment by arporter** [comment]
Changed `HaloReadAccess` and `HaloWriteAccess` so that they now check for inter-grid kernels and whether or not the field is on the fine mesh. If it is, they double the halo depth. We now get halo exchanges out to the correct depths before the two, standalone prolong and restrict examples.
Strictly speaking, we could do a `fine_field_proxy%set_clean(1)` after a prolong because we effectively compute out to L2 on the fine mesh. However, I'm not sure how useful that would be so am tempted not to worry about it in this issue.

**Comment by arporter** [comment]
I now have all tests passing, including when doing compilation. I've added `mesh_map_mod` to the stub infrastructure. Currently, `ncell_fine` is set to the last cell in the L2 halo when DM is enabled. (AFAICT, `ncell_fine` is only used to specify the extent of the second dimension of the fine dofmap.) If/when we support redundant computation in the presence of inter-grid kernels then this will need to be re-visited. Presumably, if the fields involved are discontinuous then we don't actually need to go into the halo at all (if we're not doing redundant comp)?

I missed adding inter-grid support to the kernel-stub generator off my earlier list. This feels like it could be punted to a second ticket (which I would promise to do next :-) ) but I'll let the reviewer decide.

I'll tackle tests for invokes containing multiple inter-grid kernels next...

**Comment by arporter** [comment]
I've added tests for an invoke containing multiple restrictions and prolongations and all seems fine. Halo exchanges are generated correctly. I realised that the use of "fine_" or "coarse_" in variable naming is not useful if an invoke contains cases where the same field may be viewed as both by different kernels. I've therefore dropped that qualification. This also helps with eliminating duplicate declarations/assignments.

Next step is to disable transformations on inter-grid kernels...

**Comment by arporter** [comment]
I've created #134 to tackle the issue of supporting transformations with inter-grid kernels. For this issue I've added code to reject any attempt to apply transformations to such kernels (plus associated tests).

Next stage is for me to check test coverage and @christophermaynard to check this branch (inter_grid_kernels) with LFRic...

**Comment by arporter** [comment]
@christophermaynard, @thomasmelvin I've gathered from comments in the example code that the level-1 halo of a coarse mesh is equivalent to the level-2 halo of a fine mesh. Is this always guaranteed to be the case?

**Comment by arporter** [comment]
After consultation with @christophermaynard, I have altered PSyclone to reject attempts to mix inter-grid  and non-inter-grid kernels in the same Invoke. In reviewing what I have so far, I've also taken the opportunity to handle all mesh-object related code in the same place (previously, I'd left the code that handled the `mesh` object required for DM as separate). This has meant a slight re-ordering of variable declarations in the generated output and thus I've had to update a few tests.
